test/framework/src/main/java/org/elasticsearch/cluster/ESAllocationTestCase.java

```diff
@@ -75,12 +81,42 @@ public static MockAllocationService createAllocationService(Settings settings, C
         return new MockAllocationService(
             randomAllocationDeciders(settings, clusterSettings, random),
             new TestGatewayAllocator(),
-            new BalancedShardsAllocator(settings),
+            createShardsAllocator(settings),
             EmptyClusterInfoService.INSTANCE,
             SNAPSHOT_INFO_SERVICE_WITH_NO_SHARD_SIZES
         );
     }
 
+    private static ShardsAllocator createShardsAllocator(Settings settings) {
+        return switch (SHARDS_ALLOCATOR_TYPE_SETTING.get(settings)) {
+            case BALANCED_ALLOCATOR -> new BalancedShardsAllocator(settings);
+            case DESIRED_BALANCE_ALLOCATOR -> createDesiredBalanceShardsAllocator(settings);
```

Comment:
Currently all tests rely on the default (desired balance allocator). This results in a following test failures:
```
[TrackFailedAllocationNodesTests » testTrackFailedNodes](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.TrackFailedAllocationNodesTests/testTrackFailedNodes)
[ClusterRebalanceRoutingTests » testClusterPrimariesActive1](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.ClusterRebalanceRoutingTests/testClusterPrimariesActive1)
[ClusterRebalanceRoutingTests » testAlways](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.ClusterRebalanceRoutingTests/testAlways)
[FailedShardsRoutingTests » testFailAllReplicasInitializingOnPrimaryFail](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.FailedShardsRoutingTests/testFailAllReplicasInitializingOnPrimaryFail)
[FailedShardsRoutingTests » testRebalanceFailure](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.FailedShardsRoutingTests/testRebalanceFailure)
[FailedShardsRoutingTests » testFailAllReplicasInitializingOnPrimaryFailWhil...](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.FailedShardsRoutingTests/testFailAllReplicasInitializingOnPrimaryFailWhileHavingAReplicaToElect)
[PrimaryElectionRoutingTests » testRemovingInitializingReplicasIfPrimariesFails](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.PrimaryElectionRoutingTests/testRemovingInitializingReplicasIfPrimariesFails)
[InSyncAllocationIdTests » testInSyncAllocationIdsUpdated](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.InSyncAllocationIdTests/testInSyncAllocationIdsUpdated)
[IndexBalanceTests » testBalanceIncrementallyStartNodes](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.IndexBalanceTests/testBalanceIncrementallyStartNodes)
[EnableAllocationTests » testEnableClusterBalance](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.decider.EnableAllocationTests/testEnableClusterBalance)
[RoutingIteratorTests » testNodeSelectorRouting](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.structure.RoutingIteratorTests/testNodeSelectorRouting)
[RoutingNodesIntegrityTests » testBalanceIncrementallyStartNodes](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.RoutingNodesIntegrityTests/testBalanceIncrementallyStartNodes)
[RoutingNodesIntegrityTests » testBalanceAllNodesStartedAddIndex](https://gradle-enterprise.elastic.co/s/xddpcxmjy42a2/tests/:server:test/org.elasticsearch.cluster.routing.allocation.RoutingNodesIntegrityTests/testBalanceAllNodesStartedAddIndex)
```

Reply:
Recorder above failures in https://github.com/elastic/elasticsearch/issues/86429